system/src/ledger/ledger_manager.cpp

```diff
@@ -0,0 +1,1599 @@
+    for (auto& ctx: contexts_) {
+        setPendingState(ctx.get(), PendingState::GET_INFO);
+    }
+    return 0;
+}
+
+int LedgerManager::receiveResponse(CoapMessagePtr& msg, int status) {
+    assert(state_ > State::READY);
+    auto codeClass = COAP_CODE_CLASS(status);
+    if (codeClass != 2 && codeClass != 4) { // Success 2.xx or Client Error 4.xx
+        LOG(ERROR, "Ledger request failed: %d.%02d", (int)codeClass, (int)COAP_CODE_DETAIL(status));
+        return SYSTEM_ERROR_LEDGER_REQUEST_FAILED;
+    }
+    // Get the protocol-specific result code. XXX: It's assumed that the message fields are encoded
+    // in order of their field numbers, which is not guaranteed by the Protobuf spec in general
+    char buf[32] = {};
+    int r = coap_peek_payload(msg.get(), buf, sizeof(buf), nullptr);
+    if (r < 0) {
+        if (r != SYSTEM_ERROR_END_OF_STREAM) {
+            return r;
+        }
+        r = 0; // Response is empty
+    }
+    pb_istream_t stream = pb_istream_from_buffer((const pb_byte_t*)buf, r);
+    int64_t result = 0;
+    uint32_t fieldTag = 0;
+    auto fieldType = pb_wire_type_t();
+    bool eof = false;
+    if (!pb_decode_tag(&stream, &fieldType, &fieldTag, &eof)) {
+        if (!eof) {
+            return SYSTEM_ERROR_BAD_DATA;
+        }
+        // Result code field is missing so its value is 0
+    } else if (fieldTag == PB_CLOUD(Response_result_tag)) {
+        if (fieldType != PB_WT_VARINT || !pb_decode_svarint(&stream, &result)) {
+            return SYSTEM_ERROR_BAD_DATA;
+        }
+    } // else: Result code field is missing so its value is 0
+    if (codeClass != 2) {
+        if (result == PB_CLOUD(Response_Result_OK)) {
+            // CoAP response code indicates an error but the protocol-specific result code is OK
+            return SYSTEM_ERROR_LEDGER_INVALID_RESPONSE;
+        }
+        // This is not necessarily a critical error
+        LOG(WARN, "Ledger request failed: %d.%02d", (int)codeClass, (int)COAP_CODE_DETAIL(status));
+    }
+    switch (state_) {
+    case State::SYNC_TO_CLOUD: {
+        CHECK(receiveSetDataResponse(msg, result));
+        break;
+    }
+    case State::SYNC_FROM_CLOUD: {
+        CHECK(receiveGetDataResponse(msg, result));
+        break;
+    }
+    case State::SUBSCRIBE: {
+        CHECK(receiveSubscribeResponse(msg, result));
+        break;
+    }
+    case State::GET_INFO: {
+        CHECK(receiveGetInfoResponse(msg, result));
+        break;
+    }
+    default:
+        LOG(ERROR, "Unexpected response");
+        return SYSTEM_ERROR_INTERNAL;
+    }
+    return 0;
+}
+
+int LedgerManager::receiveSetDataResponse(CoapMessagePtr& /* msg */, int result) {
+    assert(state_ == State::SYNC_TO_CLOUD && curCtx_ && curCtx_->syncDir == LEDGER_SYNC_DIRECTION_DEVICE_TO_CLOUD &&
+            curCtx_->taskRunning);
+    if (result == 0) {
+        LOG(TRACE, "Sent ledger data: %s", curCtx_->name);
+        LedgerInfo newInfo;
+        auto now = CHECK(getMillisSinceEpoch());
```

Comment:
If the RTC clock was invalid here (or in other methods in this file), would the operation get retried in a subsequent run of the state machine, or would the state machine get stuck in State::SYNC_TO_CLOUD?

Reply:
If you take a look at the CoAP callbacks that drive most of this state machine's logic, they all transition the manager to the `FAILED` state on errors and then call into the `run()` method asynchronously that will retry the sync eventually. In this case specifically, the error from `getMillisSinceEpoch()` will be propagated to `responseCallback()` which has the above logic implemented.

Reply:
Great!